modules/replicator/src/main/java/org/apache/ignite/internal/replicator/Replica.java

```diff
@@ -0,0 +1,161 @@
+     * @return Response.
+     */
+    private ReplicaResponse applyReadWriteCommandInternal(ReadWriteRequest req) {
+        //TODO: Check locks here.
+        if (req instanceof ReadWriteSingleEntryRequest) {
+            var request = (ReadWriteSingleEntryRequest) req;
+
+            var data = (SingleEntryContainer) listener.applyReadWrite(req.txId(), request.entryContainer(), req.opType());
+
+            return REPLICA_MESSAGES_FACTORY
+                    .singleEntryResponse()
+                    .partitionId(partitionId)
+                    .entryContainer(data)
+                    .build();
+        } else if (req instanceof ReadWriteMultiEntryRequest) {
+            var request = (ReadWriteMultiEntryRequest) req;
+
+            var data = (MultiEntryContainer) listener.applyReadWrite(req.txId(), request.entryContainer(), req.opType());
+
+            return REPLICA_MESSAGES_FACTORY
+                    .multiEntryResponse()
+                    .partitionId(partitionId)
+                    .entryContainer(data)
+                    .build();
+        } else {
+            throw new IgniteInternalException(
+                    IgniteStringFormatter.format("Unsupported request type [type={}]", req.getClass().getSimpleName()));
+        }
+    }
+
+    /**
+     * It is an internal method to applies a RO request to replicator.
+     *
+     * @param req Request to apply in replicator.
+     * @return Response.
+     */
+    private ReplicaResponse applyReadOnlyCommandInternal(ReadOnlyRequest req) {
+        //TODO: Check locks here.
+        if (req instanceof ReadOnlySingleEntryRequest) {
+            var request = (ReadOnlySingleEntryRequest) req;
+
+            var data = (SingleEntryContainer) listener.applyReadOnly(req.timestamp(), request.entryContainer(), req.opType());
+
+            return REPLICA_MESSAGES_FACTORY
+                    .singleEntryResponse()
+                    .partitionId(partitionId)
+                    .entryContainer(data)
+                    .build();
+        } else if (req instanceof ReadOnlyMultyEntryRequest) {
+            var request = (ReadOnlyMultyEntryRequest) req;
+
+            var data = (MultiEntryContainer) listener.applyReadOnly(req.timestamp(), request.entryContainer(), req.opType());
+
+            return REPLICA_MESSAGES_FACTORY
+                    .multiEntryResponse()
+                    .partitionId(partitionId)
+                    .entryContainer(data)
+                    .build();
+        } else {
+            throw new IgniteInternalException(
+                    IgniteStringFormatter.format("Unsupported request type [type={}]", req.getClass().getSimpleName()));
+        }
+    }
+
+    /**
+     * Process a replication request on the replica.
+     *
+     * @param request Request to replication.
+     * @return Response.
+     */
+    public ReplicaResponse processRequest(ReplicaRequest request) { // define proper set of exceptions that might be thrown.
```

Comment:
Because I think we return here to an instant response.
Probably in the future, the response will be contained a future id, by that id the future will be available.

Reply:
Nope, It's expected that we will return generally speaking the result of an async replication and wait for the future to compete only if's necessary.

Reply:
My fault, you are right.